src/ServiceControl.Monitoring/QueueLength/QueueLengthCalculator.cs

```diff
@@ -3,45 +3,81 @@
     using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
+    using System.Linq;
 
     class QueueLengthCalculator : IQueueLengthCalculator
     {
-        public void UpdateReceivedSequence(string key, long value, string queue)
+        public bool UpdateReceivedSequence(string endpointName, string key, double value, string queue)
         {
-            receivedSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
-            sessionKeyToQueueMapping.GetOrAdd(key, _ => queue);
+            var virtualQueue = new VirtualQueueId
+            {
+                EndpointName = endpointName,
+                QueueName = queue,
+                SessionKey = key
+            };
+
+            receivedSequences.AddOrUpdate(virtualQueue, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return sentSequences.ContainsKey(key);
         }
 
-        public void UpdateSentSequence(string key, long value)
+        public List<VirtualQueueId> UpdateSentSequence(string key, double value)
         {
-            sentSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
+            sentSequences.AddOrUpdate(key, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return receivedSequences.Keys.Where(k => k.SessionKey == key).ToList();
```

Comment:
Whenever `UpdateSendSequence` is called, whole `receivedSequences` concurrent dictionary is locked because of access to `Keys` collections. Is it ok?
Should we keep a separate concurrent lookup to address the query by `k.SessionKey == key`?
Should we separate the query from the command and provide another method to query by the `key`?